autoware_reference_system/test/dropped_messages.py

```diff
@@ -0,0 +1,191 @@
+# Copyright 2021 Apex.AI, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import random
+
+from bokeh.models import ColumnDataSource
+from bokeh.models import DatetimeTickFormatter
+from bokeh.models.tools import HoverTool
+from bokeh.models.widgets.tables import DataTable, TableColumn
+from bokeh.plotting import figure
+import pandas as pd
+
+
+def summary(data_model, size):
+    callback_symbols = data_model.get_callback_symbols()
+    colors = []  # Adds random colors for each callback
+    color_i = 0
+    earliest_date = None
+    latest_date = None
+    fname = ''
+    dropped_data = []
+    period_data = []
+    front_lidar_data = pd.DataFrame()
+    object_collision_data = pd.DataFrame()
+    for obj, symbol in callback_symbols.items():
+        callback_df = data_model.get_callback_durations(obj)
+        # get node information and filter out internal subscriptions
+        owner_info = data_model.get_callback_owner_info(obj)
+        if not owner_info or '/parameter_events' in owner_info:
+            continue
+        period = 0.0
+        if not owner_info or 'period' in owner_info:
+            # assume in milliseconds so convert to seconds
+            period = float(owner_info[
+                owner_info.find('period: ') + len('period: '):owner_info.rfind(' ')]) / 1000
+        if not owner_info or 'FrontLidarDriver' in owner_info and 'Timer' in owner_info:
```

Comment:
Why the `not owner_info`? Is it just to avoid an exception by `'FrontLidarDriver' in owner_info`? Is it correct to set `front_lidar_data` to `callback_df` in this case?

Reply:
> Is it correct to set front_lidar_data to callback_df in this case?

good point. @flynneva shouldn't it be `if not owner_info and` instead of `if not owner_info or`?

Reply:
I assumed the check on `owner_info` was to avoud the exception later on in case `owner_info` was `None`

Reply:
oops yes, I keep getting confused with `None` and all. Then it should be

```py
if owner_info and 'FrontLidarDriver' in owner_info and 'Timer' in owner_info:
```

or, more explicitly

```py
if owner_info is not None and 'FrontLidarDriver' in owner_info and 'Timer' in owner_info:
```

otherwise, as Ralph mentioned, if `owner_info` is `None`, you set `front_lidar_data` to `callback_df`.